c++/src/kj/mutex.c++

```diff
@@ -295,6 +298,12 @@ bool Mutex::lock(Exclusivity exclusivity, Maybe<Duration> timeout, LockSourceLoc
         state = __atomic_load_n(&futex, __ATOMIC_ACQUIRE);
       }
 
+#ifdef KJ_CONTENTION_WARNING_THRESHOLD
+      if (__atomic_exchange_n(&printContendedReader, false, __ATOMIC_RELAXED)) {
```

Comment:
This exchange is an expensive atomic requiring hardware-level locking. Can we do a double-check, i.e. check if the value is `true` first using `__atomic_load_n` (which is cheap), and only do the exchange if so?

Reply:
Would that essentially be the same thing as writing it via `atomic_compare_exchange`? I'm a little surprised that exchange is more expensive than doing a double-check although if I think about it I could see why that's the case. Is there a good tutorial you've read on the different atomic operations & the performance mental model I should have about them?

Reply:
Oh interesting. CAS is equivalent to atomic_exchange here but the double-checked lock is much faster. Not sure I understand why. https://quick-bench.com/q/2nDIxLb7d9bptd-stO03ePUh8x8

Reply:
A relaxed atomic load is "free" -- does not require any hardware-level locking. (On x86, an acquire-load is also "free", since memory ordering is guaranteed consistent.)

But any operation that has to do a read and a write together, atomically, requires hardware-level locking to make sure the read+write appears atomic to other cores.

I suppose arguably a compare-and-swap _could_ have the double-check optimization built into hardware, i.e. the CPU could do the comparison without a lock first, and skip the lock if it fails. But, that "optimization" might not be a win in most use cases, and it's easy to let software do the optimization if it wants it, so I wouldn't expect CPUs attempt such optimization.

So basically, with the double-checked approach here you avoid a hardware lock in the common case.